bundles/org.openhab.transform.rollershutterposition/src/main/java/org/openhab/transform/rollershutterposition/internal/RollerShutterPositionProfile.java

```diff
@@ -240,40 +231,109 @@ public void run() {
             }
 
             logger.trace("stopTimeoutTask() position: {}", targetPosition);
-
-            if (updateTimer != null) {
-                Objects.requireNonNull(updateTimer).cancel(true);
-                updateTimer = null;
-            }
-
+            position = currentPosition();
+            callback.sendUpdate(new PercentType(position));
             movingSince = Instant.MIN;
-            position = targetPosition;
             targetPosition = -1;
-            callback.sendUpdate(new PercentType(position));
         }
     };
 
     // Runnable task to update the item on position while the roller shutter is moving
     private Runnable updateTimeoutTask = new Runnable() {
         @Override
         public void run() {
+            logger.trace("updateTimeoutTask() called, isMoving: {}, currentPosition: {}", isMoving(),
+                    currentPosition());
             if (isMoving()) {
                 int pos = currentPosition();
-                if (pos < 0 || pos > 100) {
-                    return;
+                if ((pos <= 0 && direction == UpDownType.UP) || (pos >= 100 && direction == UpDownType.DOWN)) {
+                    logger.debug("updateTimeoutTask() reached end position, stopping update timer");
+                    stopUpdateTimer();
                 }
+                pos = Math.max(0, Math.min(100, pos));
                 callback.sendUpdate(new PercentType(pos));
-                logger.trace("updateTimeoutTask(): {}", pos);
             }
         }
     };
 
+    @Override
+    public void onCommandFromItem(Command command) {
+        logger.debug("onCommandFromItem: {}", command);
+
+        // pass through command if profile has not been configured properly
+        if (!isValidConfiguration) {
+            callback.handleCommand(command);
+            return;
+        }
+
+        if (command instanceof UpDownType) {
+            if (command == UpDownType.UP) {
+                moveTo(0);
+            } else if (command == UpDownType.DOWN) {
+                moveTo(100);
+            }
+        } else if (command instanceof StopMoveType) {
+            stop(false);
+        } else {
+            moveTo(((PercentType) command).intValue());
```

Comment:
As there is no check anywhere for the type, it is risky to do this. Maybe the current codepath would not allow anything else, but future refactoring might introduce issues. Better to explicitly check with instanceof as above. Leave the else to a logger that the  command cannot be processed or similar.

```suggestion
            moveTo(((PercentType) command).intValue());
```